### DISCUSSION TITLE
How to use _environment in code chunk?

### DISCUSSION THREAD

**Discussion by AndhikaWB** [quarto-dev/quarto-cli#10448] [Q&A] [answered]
### Description

I have set a secret variable in `_environment` file and also created a basic `_quarto.yml` file on my project

However, I can't find a way to use the environment written here in my Quarto notebook. For example, here's my code chunk, where `BQ_EMAIL` is set in `_environment`:
```
{r setup-sql-conn}

library(DBI)
library(bigrquery)

cred_file <- paste0(
  'docker/root/.config/gcloud/legacy_credentials/',
  Sys.getenv(BQ_EMAIL), '/adc.json'
)

Sys.setenv(GOOGLE_APPLICATION_CREDENTIALS = cred_file)
bq_auth()

db <- dbConnect(
  bigquery(), 
  dataset = 'sales_data', 
  project = 'mumu-431300', 
  use_legacy_sql = FALSE
)
```

It should be recognized by Quarto and the R chunk, but it's not, the output is:
```
Error in eval(expr, envir, enclos): object 'BQ_EMAIL' not found
```

The docs written [here](https://quarto.org/docs/projects/environment.html) doesn't explain a way to use the environment either, only how to set it. So how do I really use it? Using `{{{< env BQ_EMAIL >}}}` doesn't seem to work either (in code chunk)

**Comment by mcanouil** [comment]
Could you share a small reproducible example of your setup?
Also which version of Quarto are you using?

Do you have a `_quarto.yml` as documented? (https://quarto.org/docs/projects/environment.html#environment-file)
I suspect you don't.

> If you include an `_environment` file at the root level of your project alongside `_quarto.yml` the values defined within the file will be read and set into the environment when the project configuration is read.

**Reply by AndhikaWB** [reply]
@mcanouil thank you for replying. I'm using Quarto v1.5.54

Here is my `_quarto.yml` file (not sure what the bare minimum settings is, not documented well):
```
project:
  type: default
```

Here is my `_environment` file:
```
BQ_EMAIL=example@gmail.com
```

Both are in the same folder, with structure like this (`notebook.qmd` is my main file):
![image](https://github.com/user-attachments/assets/03523065-86b6-4b20-95c5-aaff9de4767c)

Here is the header of my `notebook.qmd`, with the important code in my previous post:
```
---
title: BigQuery Test
author: AndhikaWB
date: last-modified
date-format: long
format:
  gfm:
    output-file: README.md
knitr:
  opts_chunk:
    echo: true
    error: true
    cache: true
    results: hold
---
```

I have tried executing `quarto render` manually (without any extra parameter) and by clicking Quarto preview button on VS Code

**Reply by mcanouil** [reply, marked as answer]
Unfortunately your examples are not reproducible.
Below a screenshot of a simple setup following the documentation:
<img width="1624" alt="image" src="https://github.com/user-attachments/assets/1f479bfd-c49f-4c61-a922-1a40360c8289">

**Reply by GuillaumeChretien** [reply]
Hello, I also had difficulties finding this simple line. Is it possible to suggest that this example be added here https://quarto.org/docs/projects/environment.html? Best regards.

**Reply by AndhikaWB** [reply]
Sorry again for troubling you mcanouil 

It seems that I forgot to add surrounding quotes so I wrote `Sys.getenv(BQ_EMAIL)` instead of `Sys.getenv("BQ_EMAIL")`. The issue isn't about Quarto _environment itself

I think I need to take a break this weekend since I made too many mistakes lately 😅

**Reply by mcanouil** [reply]
@GuillaumeChretien What "simple line"?

**Comment by GuillaumeChretien** [comment]
The simple Sys.getenv(...) in a r chunk (and with an ojs_define it us very useful !). I thought it could be efficient to document it in https://quarto.org/docs/projects/environment.html

**Reply by mcanouil** [reply]
I don't think that's up to Quarto to document how to access environment variables from every single languages supported by knitr or Jupyter.
This information is usually easily accessible/searchable in the documentation of each languages.

**Reply by GuillaumeChretien** [reply]
Thank you for your response. It seems I am missing the documentation on this specific point because I have only found a few hints in discussion forums.

**Reply by mcanouil** [reply]
I am not sure to understand.

The environment variable page is there to document that it is possible to set environment variables that will live in all steps of Quarto: pre/post render scripts, LUA filter/shortcode, code cells, env shortcode, etc.

How to use those environment variables is 100% up to the users as the use cases are numerous.

If you don't know how to use environment variables in bash, R, Python, etc. that's not something that Quarto can teach you about as Quarto tries to be agnostic of the labguages and this is highly language dependent.

**Reply by GuillaumeChretien** [reply]
Thank you, I understand the logic well, especially since I navigate between R, Python, and OJS as needed. It was just modest user feedback: a simple example like the one you presented earlier would have helped me with reading the documentation on environment variables. Again, thank you for all your insights.
